src/main/java/graphql/schema/SchemaUtil.java

```diff
@@ -98,6 +101,8 @@
 
     void replaceTypeReferences(GraphQLSchema schema) {
         final Map<String, GraphQLType> typeMap = schema.getTypeMap();
-        TRAVERSER.depthFirst(new GraphQLTypeResolvingVisitor(typeMap), typeMap.values());
+        List<GraphQLType> roots = new ArrayList<>(typeMap.values());
+        roots.addAll(schema.getDirectives());
```

Comment:
`GraphQLArgument` has previously only been found on a `GraphQLFieldDefinition` which itself could only been found on a `GraphQLObjectType` which is regular type. As a result, a type reference in an argument would only be possible within a hierarchy of another type. Now, a `GraphQLArgument` can be found directly on a directive which is not a real type. As a result, a type reference on an argument can be found outside of a hierarchy of a real type, thus it will not be in the schema's type map. And for those to be replaced correctly, directives need to be added as roots. This may result in a duplicated traversal in some cases (if the type of the directive argument is complex and has a reference inside, and _does_ end up in the type map), but outside of a small performance hit, it has no negative consequences.

Reply:
I see what you are proposing, but not sure I have good enough picture about possible use and edge cases.

Do we allow directives with same names and different arguments?
Can I have a directive, which accepts a complex type which has yet another directive and another type? 
Will it matter to distinguish type traversal from traversal of a type under directive(s)?

Please bear with my questions, I will try to catch up with directives and clear my vision.
So far most simplest solution is probably the best strategy :)

Reply:
But what I think is worth pointing out is that I don't change any rules in this PR. Everything doable with my changes is also doable without them. You can already add directives to anything, assign arguments of complex types, add further directives to those arguments etc.

All that the change does is make directives reachable by the traverser and, using that, make sure the reference replacement and type collection work in a consistent manner.